Support out of box adding lineItems for all open invoice methods (#918)
commercetools/commercetools-adyen-integration: docs/adr/0005-adding-line-items.md

# 4. Notification error handling.

Date: 2022-02-01

## Status

[On Review]()

## Context

- Many payment methods requires information about the price and product information by specifying `lineItems` in `makePaymentRequest`.

> - Integration supports automatically **adding lineItems** from commercetools cart for [klarna and affirm payments](https://github.com/commercetools/commercetools-adyen-integration/blob/v9.6.0/extension/docs/WebComponentsIntegrationGuide.md#klarna-payment-and-affirm-payment).

Reviewer: ```suggestion
- Integration can **add lineItems** from commercetools cart for [klarna and affirm payments](https://github.com/commercetools/commercetools-adyen-integration/blob/v9.6.0/extension/docs/WebComponentsIntegrationGuide.md#klarna-payment-and-affirm-payment).
```

> - There are many payment methods which require line items **beside** the klarna and affirm payment.

Reviewer: ```suggestion
- There are many payment methods which require line items **besides** the klarna and affirm payment.
```

> - Since we add out of the box for Affirm and Klarna only the users need to write additional logic to include those items for other payment methods which poses an inconvenience. For instance check the request here on issue [#728](https://github.com/commercetools/commercetools-adyen-integration/issues/728).

Reviewer: ```suggestion
- Integration adds lineItems for Affirm and Klarna only. Users need to write additional logic to include those lineItems for other payment methods which poses an inconvenience. For instance check the request here on issue [#728](https://github.com/commercetools/commercetools-adyen-integration/issues/728).
```

## Decision

- If lineItems field is provided with `makePaymentRequest`, extension will skip adding lineItems and leave the provided lineItems.

> - An application configuration flag `addCommercetoolsLineItems: true/false` for extension, which will default to false and if set to true would add lineItems to all payment methods.

Reviewer: ```suggestion
- An application configuration flag `addCommercetoolsLineItems: true/false` for extension, which will default to **false**. If set to **true**, it would add lineItems to all payment methods.
```

> - An application configuration flag `addCommercetoolsLineItems: true/false` for extension, which will default to **false**. If set to **true**, it would add lineItems to payment methods that requires lineItems.

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/918#discussion_r800378868

> - If user includes in the makePaymentRequest,  "addCommercetoolsLineItems": true then integration would generate line items, no matter if it's set to false on application config.

Author: is it something like forcing line item generation for other payment methods like credit card ? 🤔

Reviewer: So this point: 
```
Extension does not include lineItems for payment methods where those infos are not beneficial and thus we save an API call i.e: fetch the cart.
```
Will it work if the payment method is not the one that requires [lineItems]?

Author: Yes it works, but not required. I guess those add additional fields on response related to fraud calculation, so does not seem harmful.

https://github.com/commercetools/commercetools-adyen-integration/pull/918/commits/bae78ed26e44a24a3c49de3412182d4358ec40c2#diff-e2e07d6c2da8d9a9bdf7f434ef5c8c01b699a20c81feb1d379dbc12317afd5caR44-R48

> - If user includes in the makePaymentRequest,  "addCommercetoolsLineItems": true then integration would generate line items, no matter if it's set to false on application config.

Reviewer: ```suggestion
- If user sets  `"addCommercetoolsLineItems": true` in the `makePaymentRequest`, then integration would generate line items, regardless of the setting in the application config.
```

> - For backwards compatibility, still add line items for both Affirm and Klarna. No matter if "addCommercetoolsLineItems" is set to true or false.

Reviewer: ```suggestion
- For backwards compatibility, the integration still adds line items for both Affirm and Klarna. No matter if `addCommercetoolsLineItems` is set to true or false.
```

> - Extension does not include lineItems for payment methods where those infos are not beneficial and thus we save an API call i.e: fetch the cart.

Reviewer: ```suggestion
- Extension will not add lineItems for payment methods where lineItems are not beneficial and thus we save an API call to fetch the cart.
```

Payment method types that requires [lineItems](https://docs.adyen.com/api-explorer/#/CheckoutService/latest/payments__reqParam_lineItems): 

> `afterpay`, `afterpaytouch`, `klarna`, `ratepay`, `facilypay`, `clearpay`, `grabpay`, `paybright`, `pix`, `zip`.

Author: I collect the list with checking the web component docs one by one, similar question asked by Lam before: https://github.com/commercetools/commercetools-adyen-integration/pull/918#discussion_r799208645

3x 4x oney -> facilypay on payment type.
Atome - probably grabpay, could not find much information on the list.
